### ISSUE TITLE
Dirac Error: 'str' object has no attribute 'get'

### ISSUE THREAD

**Issue by nasfarley88** [ganga-devs/ganga#550]
Starting at v601r20, I get the following error 

```
ERROR    Dirac Error: 'str' object has no attribute 'get'
INFO     submitting job 128
INFO     job 128 status changed to "submitting"
INFO     Job 128: Preparing DaVinci application.
INFO     submitting job 129
INFO     job 129 status changed to "submitting"
INFO     Job 129: Preparing DaVinci application.
INFO     Splitting Job: 126
INFO     Constructing subjobs
INFO     Requesting LFN replica info
ERROR    Dirac Error: 'str' object has no attribute 'get'
ERROR    Dirac Error: 'str' object has no attribute 'get'
ERROR    Dirac Error: 'str' object has no attribute 'get'
ERROR    Dirac Error: 'str' object has no attribute 'get'

Traceback (most recent call last):
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/Core/GangaThread/WorkerThreads/WorkerThreadPool.py", line 114, in __worker_thread
    result = item.command_input.function(*these_args, **item.command_input.kwargs)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaDirac/Lib/Splitters/OfflineGangaDiracSplitter.py", line 86, in getLFNReplicas
    str(this_min), str(this_max), len(allLFNs)))
TypeError: 'str' object is not callable

ERROR    Traceback (most recent call last):
  File "<stdin>", line 422, in <module>
  File "<string>", line 8, in <module>
ImportError: No module named DIRAC.Core.Base.Script

ERROR    Unknown error ion Dirac LFN Failed output
ERROR    'str' object has no attribute 'get' ... reverting job 126 to the new status
WARNING  ---------- error in user/extension code ----------
WARNING  Traceback (most recent call last):
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1574, in submit
WARNING      rjobs = self._doSplitting()
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1448, in _doSplitting
WARNING      subjobs = self.splitter.validatedSplit(self)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Adapters/ISplitter.py", line 70, in validatedSplit
WARNING      subjobs = self.split(stripProxy(job))
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaLHCb/Lib/Splitters/SplitByFiles.py", line 195, in split
WARNING      split_return = super(SplitByFiles, self).split(job)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaGaudi/Lib/Splitters/GaudiInputDataSplitter.py", line 70, in split
WARNING      for dataset in all_jobs:
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaDirac/Lib/Splitters/OfflineGangaDiracSplitter.py", line 355, in OfflineGangaDiracSplitter
WARNING      errors = sortLFNreplicas(bad_lfns, allLFNs, LFNdict, ignoremissing, allLFNData, inputs)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaDirac/Lib/Splitters/OfflineGangaDiracSplitter.py", line 218, in sortLFNreplicas
WARNING      raise err
WARNING  AttributeError: 'str' object has no attribute 'get'
WARNING  --------------------------------------------------
ERROR    Exception raised executing 'submit' in Thread 'None':
Traceback (most recent call last):
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/Core/GangaThread/WorkerThreads/WorkerThreadPool.py", line 114, in __worker_thread
    result = item.command_input.function(*these_args, **item.command_input.kwargs)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Base/Proxy.py", line 642, in proxy_wrapped
    r = f(*s_args, **s_kwargs)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1574, in submit
    rjobs = self._doSplitting()
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1448, in _doSplitting
    subjobs = self.splitter.validatedSplit(self)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/Ganga/GPIDev/Adapters/ISplitter.py", line 70, in validatedSplit
    subjobs = self.split(stripProxy(job))
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaLHCb/Lib/Splitters/SplitByFiles.py", line 195, in split
    split_return = super(SplitByFiles, self).split(job)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaGaudi/Lib/Splitters/GaudiInputDataSplitter.py", line 70, in split
    for dataset in all_jobs:
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaDirac/Lib/Splitters/OfflineGangaDiracSplitter.py", line 355, in OfflineGangaDiracSplitter
    errors = sortLFNreplicas(bad_lfns, allLFNs, LFNdict, ignoremissing, allLFNData, inputs)
  File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v601r20/install/ganga/python/GangaDirac/Lib/Splitters/OfflineGangaDiracSplitter.py", line 218, in sortLFNreplicas
    raise err
JobError: JobError: Error: 'str' object has no attribute 'get'
```

but with v601r19 I get no error. 

More details:
https://lhcbqa.web.cern.ch/lhcbqa/172/dirac-error-str-object-has-no-attribute-get

Skeleton job submission script:
``` python
 j = Job()

 j.application = DaVinci(version="v37r2p2")

 j.splitter = SplitByFiles(filesPerJob=100) # default 10, max 100

 j.do_auto_resubmit = True

 j.backend=Dirac()

 j.inputdata = j.application.readInputData('/path/to/data/file.py')
 j.application.optsfile = '/path/to/opts/file.py'
 j.application.extraopts = ";".join([
     "from Configurables import DaVinci",
     "DaVinci().DataType = \"2011\"",
     "DaVinci().Lumi = False",
     "DaVinci().Simulation = True",
 ])
 j.comment = "file.root"
+j.outputfiles = [DiracFile('davinci.root')]

 queues.add(j.submit)
```
I'm happy to run things if you have trouble reproducing the bug :)

**Comment by rob-c** [comment]
@nasfarley88 You are hitting a strange bug which appears to be effecting a subset of users.

The main error from this is: `ImportError: No module named DIRAC.Core.Base.Script`

This is caused by Ganga (for some reason that we're yet to determine) failing to source the LHCbDirac environment successfully.
The way that this code works has completely changed in the build up to 6.1.21 as to avoid this potential problem in the future.

I am occasionally able to regenerate this, but completely unreliably in 6.1.20. It's rare enough that it didn't show up in release testing.
I have found that switching lxplus node and sometimes running: `mv ~/.cache/LHCbDIRAC_ENV/ ~/.cache/LHCbDIRAC_ENV_bak ; SetupProject LHCbDirac ; SetupProject Ganga v601r19; SetupProject Ganga v601r20` helps as this forces the LHCbDirac environment to be regenerated.

I'm going to keep this bug open to see if an obvious workaround can be spotted. Either way this has already been fixed ahead of 6.1.21 which should be released within a week or 2 I hope.

I have a suspicion that this could be a symptom of some users who are using tcsh? If you're not using bash as your default shell I'd be curious to know as the code in 6.1.20 calls `/bin/sh` but the code in 6.1.21 calls `bash` as our code is expecting this to be the default shell it's executed in.

**Comment by nasfarley88** [comment]
Interesting, I'm running from cvmfs and using zsh. Are there tests that can be run on my end to see whether these things are the problem?

**Comment by rob-c** [comment]
@nasfarley88 If you were to drop into a bash session by running `bash` (I don't think you should need to pass it any arguments) and launch ganga after running SetupProject are you able to submit your job successfully?

**Comment by nasfarley88** [comment]
@rob-c I've dropped into bash and run the string of commands you recommended before and it seems to be working. When my job finishes I'll test with just `SetupProject Ganga`.

**Comment by rob-c** [comment]
@nasfarley88 OK, good to know this issue is caused by users launching Ganga in a different shell.

Would you feel confident at installing Ganga locally and trying to run against our latest develop branch to see if you still have problems?

If not, I'll try and run a test of submitting jobs from within the zsh and tclsh environments just to be sure we're not breaking anything in the future.

Thanks,

Rob

**Comment by drmarkwslater** [comment]
@rob-c Just spoken to @nasfarley88 - after his jobs have finished he'll try running off CVMFS using lb-run which has the latest develop available.

**Comment by nasfarley88** [comment]
FYI, just ran SetupProject Ganga in bash without the `mv`... commands, and it couldn't import the `DIRAC.Core.Base.Script` module. 

I'll run the dev version once CVMFS catches it.

**Comment by rob-c** [comment]
This should be gone in 6.1.21 due to the changes in reading in the config files. The workaround here works as discussed so I think we can close this once 6.1.21 is out and recommend people update to that version.

**Comment by rob-c** [comment]
Fixed in 6.1.22 which will be released shortly
